docs/config.md
# Configuration File

Below is a list of the available configuration options for Revizor, which are passed down to Revizor via a config file.
For an example of how to write the config file, see [src/tests/big-fuzz.yaml](https://github.com/microsoft/sca-fuzzer/tree/main/src/tests/big-fuzz.yaml).

## Fuzzing Configuration

```yaml
Name: fuzzer
Default: 'basic'
Options: 'basic' | 'architectural' | 'archdiff'
```

This option selects the fuzzing mode. The available options are:

* `basic` - normal model-based fuzzing. A violation in this mode indicates that the CPU
exposes more information than predicted by the contract. This option should be used in most
testing campaigns.
* `architectural` - self-fuzzing for architectural mismatches between the model and the executor.
This option should be used for testing the fuzzer itself, i.e., a violation in this
mode indicates a bug in the fuzzer rather then a bug in the CPU. This is useful when running
the fuzzer with a previously-untested instruction set, or when a new contract is implemented.
* `archdiff` - fuzzing for architectural invariants. This is a special mode targeted for
for semi-microarchitectural violations, similar to ZenBleed. This mode is experimental and
should be used with caution.

```yaml
Name: enable_priming
Default: True
```

This option enables or disables priming. This options should be set to True in most cases,
as priming is crucial for eliminating false positives.

Priming solves the following problem: Revizor collects hardware traces for inputs in a sequence,
and the microarchitectural state is not reset between the inputs. This means that the microarchitectural
state for the input at, for example, position 100 is different from the state for the input at position 200.
Accordingly, the hardware traces for these inputs may differ because the measurements are taken in different
microarchitectural contexts.

To address this issue, we uses priming, which swaps the inputs in the sequence and re-runs the tests.
For example, if the original sequence is (i1 . . . i99,i100,i101 . . . i199,i200), the priming
sequence will be (i1 . . . i99,i200,i101 . . . i199,i100). If the violation persists in this
sequence, it is a true positive. If the violation disappears, it is a false positive, and it
will be discarded.

```yaml
Name: enable_speculation_filter
Default: False
```

If enabled, Revizor will discard test cases that do not trigger speculation.

This option is useful for improving the throughput of the fuzzer,
but it can discard potential violations if the leakage is not caused by speculation.

```yaml
Name: enable_observation_filter
Default: False
```

If enabled, Revizor will discard test cases that do not leave speculative traces.
The filtering is performed by adding an `LFENCE` after
each instruction in the test case, and comparing the resulting hardware traces with the original.
If the traces are identical, the test case is discarded.

This option is useful for improving the throughput of the fuzzer,
but it can discard potential violations if the leakage is not caused by speculation.

```yaml
Name: enable_fast_path_model
Default: True
```

If enabled, the same contract trace will be used for all inputs in the same taint-based input class.

```yaml
Name: color
Default: False
```

If enabled, the output will be colored.
This option is helps a lot with readability, but may produce corrupted output when redirected to a file.

```yaml
Name: logging_modes
Default: ['info', 'stat']
Options: 'info' | 'stat' | 'dbg_timestamp' | 'dbg_violation' | 'dbg_dump_htraces' | 'dbg_dump_ctraces' | 'dbg_dump_traces_unlimited' | 'dbg_executor_raw' | 'dbg_model' | 'dbg_coverage' | 'dbg_generator' | 'dbg_priming'
```

This option controls the output:

* `info` - general information about the progress of fuzzing;
* `stat` - statistics the end of the fuzzing campaign;
* `dbg_timestamp` - every 1000 test cases print the timestamp during the fuzzing process;
* `dbg_violation` - upon detecting a violation, print detailed information about it;
* `dbg_dump_htraces` - print the first 100 hardware traces for every test case;
* `dbg_dump_ctraces` - print the first 100 contract traces for every test case;
* `dbg_dump_traces_unlimited` - print ALL traces (use carefully, produces LOTS of text);
* `dbg_executor_raw` - prints hardware traces for every stage of the fuzzing process;
  this differs from `dbg_dump_htraces` in that it prints the traces collected by
  speculation/observation filters as well as at every iteration of multi-sample collection;
* `dbg_model` - print a detailed info about EVERY instruction executed on the model (use carefully, produces LOTS of text);
* `dbg_coverage` - stores instruction coverage information;
* `dbg_generator` - prints a list of instructions used to generate test cases;
* `dbg_priming` - prints information about the priming process; only useful for debugging the priming mechanism itself.

```yaml
Name: multiline_output
Default: False
```

If enabled, each output message will be printed on a separate line.
Otherwise, the fuzzing progress will be continuously overwriting the same line (works only in the terminal).


## Program Generator Configuration

```yaml
Name: instruction_set
Default: 'x86-64'
Options: 'x86-64'
```

The instruction set under test. Currently, only x86-64 is supported.

```yaml
Name: instruction_categories
Default: ['BASE-BINARY', 'BASE-BITBYTE', 'BASE-COND_BR']
Options: 'BASE-BINARY' | 'BASE-BITBYTE' | 'BASE-CMOV' | 'BASE-COND_BR' | 'BASE-CONVERT' | 'BASE-DATAXFER' | 'BASE-FLAGOP' | 'BASE-LOGICAL' | 'BASE-MISC' | 'BASE-NOP' | 'BASE-POP' | 'BASE-PUSH' | 'BASE-SEMAPHORE' | 'BASE-SETCC' | 'BASE-STRINGOP' | 'BASE-WIDENOP' | 'BASE-INTERRUPT' | 'BASE-SYSTEM' | 'LONGMODE-CONVERT' | 'LONGMODE-DATAXFER' | 'LONGMODE-SEMAPHORE' | 'LONGMODE-SYSCALL' | 'LONGMODE-SYSRET' | 'SSE-SSE' | 'SSE-DATAXFER' | 'SSE-MISC' | 'SSE-LOGICAL_FP' | 'SSE2-SSE' | 'SSE2-DATAXFER' | 'SSE2-MISC' | 'SSE2-LOGICAL_FP' | 'SSE2-LOGICAL' | 'SSE3-SSE' | 'SSE3-DATAXFER' | 'SSE4-LOGICAL' | 'SSE4a-BITBYTE' | 'SSE4a-DATAXFER' | 'CLFLUSHOPT-CLFLUSHOPT' | 'CLFSH-MISC' | 'MPX-MPX' | 'SMX-SYSTEM' | 'VTX-VTX' | 'XSAVE-XSAVE'
```

Select a list of instruction categories to be used when generating programs.
This list effectively filters out instructions from the ISA descriptor file (e.g., `base.json`)
passed via the command line (`-s`).

```yaml
Name: instruction_blocklist
Default: ['enterw', 'enter', 'leavew', 'leave', 'int', 'encls', 'vmxon', 'stgi', 'skinit', 'ldmxcsr', 'stmxcsr', 'lfence', 'mfence', 'sfence', 'clflush', 'clflushopt', 'divps', 'divss', 'divpd', 'divsd', 'mulss', 'mulps', 'mulpd', 'mulsd', 'rsqrtps', 'rsqrtss', 'sqrtps', 'sqrtss', 'sqrtpd', 'sqrtsd', 'addps', 'addss', 'addpd', 'addsd', 'subps', 'subss', 'subpd', 'subsd', 'addsubpd', 'addsubps', 'haddpd', 'haddps', 'hsubpd', 'hsubps', 'sti', 'cli', 'xlat', 'xlatb', 'cmpxchg8b', 'lock cmpxchg8b', 'cmpxchg16b', 'lock cmpxchg16b', 'cpuid', 'cmpps', 'cmpss', 'cmppd', 'cmpsd', 'movq2dq', 'movdq2q', 'rcpps', 'rcpss', 'maskmovdqu']
Options: (any instruction names)
```

A list of instructions that will NOT be used for generating programs.
This list filters out instructions from `instruction_categories`, but not from `instruction_allowlist`.

The resulting instruction pool is:
     (instructions from instruction_categories - instruction_blocklist) + instruction_allowlist

The instructions that are blocked by default are known to cause issues in the model or executor,
and hence should generally be avoided when fuzzing.

```yaml
Name: instruction_blocklist_append
Default: []
Options: (any instruction names)
```

A list of instructions that will be appended to the default instruction blocklist.
This option is identical to `instruction_blocklist`, but the list is added to the previous
blocklist instead of replacing it.
This is useful when you want to block some instructions in addition to the default blocklist.

```yaml
Name: instruction_allowlist
Default: []
Options: (any instruction names)
```

A list of instructions to use for generating programs.
This list has priority over `instruction_categories` and over `instruction_blocklist`,
thus adding instructions on top of the categories.

The resulting instruction pool is:
     (instructions from instruction_categories - instruction_blocklist) + instruction_allowlist

```yaml
Name: program_generator_seed
Default: 0
```

Seed of the program generator. If set to zero, a random seed will be used for each run.


```yaml
Name: program_size
Default: 24
```

Number of instructions per program. The actual size might be larger because of the instrumentation.


```yaml
Name: avg_mem_accesses
Default: 12
```

Average number of memory accesses in generated programs.
The actual number will be random, but the average over all programs will be close to this value.

```yaml
Name: min_bb_per_function
Default: 1
```

Minimal number of basic blocks per function in generated programs.

```yaml
Name: max_bb_per_function
Default: 2
```

Maximal number of basic blocks per function in generated programs.

```yaml
Name: min_successors_per_bb
Default: 1
```

Minimal number of successors for each basic block in generated programs.

Note 1: this config option is a *hint*; it could be ignored if the instruction set does not
have the necessary instructions to satisfy it, or if a certain number of successor is required
for correctness

Note 2: If min_successors_per_bb > max_successors_per_bb, the value is
overwritten with max_successors_per_bb

```yaml
Name: max_successors_per_bb
Default: 1
```

Maximal number of successors for each basic block in generated programs.

Note: this config option is a *hint*; it could be ignored if the instruction set does not
have the necessary instructions to satisfy it, or if a certain number of successor is required
for correctness

```yaml
Name: register_allowlist
Default: []
Options: (any register names)
```

A list of registers that CAN be used for generating programs.

This list has higher priority than `register_blocklist`.
The resulting list is: (all registers - `register_blocklist`) + `register_allowlist`.

```yaml
Name: register_blocklist
Default: (all but RAX, RBX, RCX, RDX, RDI, RSI, XMM0-XMM7)
Options: (any register names)
```

A list of registers that will NOT be used for generating programs.

This list has lower priority than `register_allowlist`.
The resulting list is: (all registers - `register_blocklist`) + `register_allowlist`.

The default blocked registers are used by the executor internally, and thus should be avoided.

```yaml
Name: generator_faults_allowlist
Default: []
Options: 'div-by-zero' | 'div-overflow' | 'opcode-undefined' | 'bounds-range-exceeded' | 'breakpoint' | 'debug-register' | 'non-canonical-access' | 'user-to-kernel-access'
```

By default, the generator will produce programs that never trigger exceptions.
This option modifies this behavior by permitting the generator to produce 'unsafe' instruction sequences
that could potentially trigger an exception. The model and executor will also be configured to handle
these exceptions gracefully.

The available options are:

* `div-by-zero` - generate divisions with unmasked divisor, which can cause a division by zero exception.
* `div-overflow` - generate divisions with unmasked dividend, which can cause an overflow exception.
* `opcode-undefined` - generate undefined opcodes, which can cause an undefined opcode exception.
* `bounds-range-exceeded` - apply MPX instructions for random bounds checks.
  This is possible only if MPX is included in the tested instruction set.
* `breakpoint` - generate breakpoints, which can cause INT3 exceptions.
* `debug-register` - generate instructions that cause INT1 exceptions.
* `non-canonical-access` - randomly select a memory access in a generated program and instrument it to access a non-canonical address.
* `user-to-kernel-access` - randomly select a memory access in a user actor's code and instrument it to access a kernel address; works only if there is at least one user actor.

## Actor Configuration

All actors are defined in the `actors` list, with the following syntax:

```yaml
actors:
  - <actor1_name>
    - <actor_option>: <value>
    - <actor_option>:
       - <sub_option1>: <value1>
       - <sub_option2>: <value2>
    ...
  - <actor2_name>
      ...
  ...
```

The following options are available for each actor:

```yaml
Actor Option: mode
Default: 'host'
Options: 'host' | 'guest'
```

The execution mode of the actor. The available options are:

* `host` - the actor runs in the normal, non-virtualized mode.
* `guest` - the actor runs in a VM (one VM per actor).

```yaml
Actor Option: privilege_level
Default: 'kernel'
Options: 'user' | 'kernel'
```

The privilege level of the actor. The available options are:

* `user` - the actor runs in user mode (CPL=3).
* `kernel` - the actor runs in kernel mode (CPL=0).

```yaml
Actor Option: data_properties
Default: (see below)
Options: 'present' | 'writable' | 'user' | 'accessed'
         | 'dirty' | 'executable' | 'reserved_bit' | 'randomized'
```

The properties of the data memory used by the actor.
These properties are applied only to the second page (FAULTY_AREA) of the actor's data region.

The available options are:

* `present` [default: True] - the value of the Present bit in the page table entry.
* `writable` [default: True] - the value of the Writable bit in the page table entry.
* `user` [default: False] - the value of the User/Supervisor bit in the page table entry.
* `accessed` [default: True] - the value of the Accessed bit in the page table entry.
* `dirty` [default: True] - the value of the Dirty bit in the page table entry.
* `executable` [default: False] - the value of the Executable bit in the page table entry.
* `reserved_bit` [default: False] - the value of the Reserved bit in the page table entry.
* `randomized` [default: False] - if true, the values of the above properties will be randomized for each test case.

Note that the above properties are set in the host page tables for actors with `mode: host`,
and in the guest page tables for actors with `mode: guest`.

```yaml
Actor Option: data_ept_properties
Default: (see below)
Options: 'present' | 'writable' | 'executable' | 'accessed' | 'dirty' | 'user'
        | 'reserved_bit' | 'randomized'
```

The properties of the EPT entry used by the actor (on Intel) or the NPT entry (on AMD).
The properties are applied only to the second page (FAULTY_AREA) of the actor's data region.

This property has no effect on actors with `mode: host`.

The available options are:


src/config.py
"""
File: Fuzzing Configuration Options

Copyright (C) Microsoft Corporation
SPDX-License-Identifier: MIT
"""
import yaml
import os
from copy import deepcopy
from typing import List, Dict, IO, Any
from collections import OrderedDict
from .x86 import x86_config


# ==================================================================================================
# Helper classes
# ==================================================================================================
class IncludeLoader(yaml.SafeLoader):
    """
    Helper class to enable `!include` statements in configuration files
    """
    visited: List[str] = []

    def __init__(self, stream: IO, include_dir: str = "") -> None:
        self._search_paths = [os.path.split(stream.name)[0]]
        if include_dir:
            self._search_paths.append(include_dir)
        self.visited.append(os.path.abspath(stream.name))
        super(IncludeLoader, self).__init__(stream)

    def __del__(self) -> None:
        if self.visited:
            self.visited.pop()

    def include(self, node: yaml.Node) -> Any:
        """
        Include another YAML file
        """
        # find the included file
        for root in self._search_paths:
            filename = os.path.join(root, self.construct_scalar(node))  # type: ignore
            if os.path.exists(filename):
                break
        else:
            raise ConfigException(f"Included file {filename} does not exist")

        # check for cycles
        if os.path.abspath(filename) in self.visited:
            raise ConfigException(f"Circular include detected in {filename}")

        with open(filename, 'r') as f:
            return yaml.load(f, IncludeLoader)


IncludeLoader.add_constructor('!include', IncludeLoader.include)


class ConfigException(SystemExit):

    def __init__(self, message: str) -> None:
        super().__init__("\nCONFIG ERROR: " + message + "\n")


# ==================================================================================================
# Main configuration class
# ==================================================================================================
class Conf:
    # ==============================================================================================
    # Fuzzer
    fuzzer: str = "basic"
    """ fuzzer: type of the fuzzing algorithm """
    enable_priming: bool = True
    """ enable_priming: whether to check violations with priming """
    enable_speculation_filter: bool = False
    """ enable_speculation_filter: if True, discard test cases that don't trigger speculation"""
    enable_observation_filter: bool = False
    """ enable_observation_filter: if True,discard test cases that don't leave speculative traces"""
    enable_fast_path_model: bool = True
    """ enable_fast_path_boosting: if enabled, the same contract trace will be used
    for all inputs in the same taint-based input class """

    # ==============================================================================================
    # Program Generator
    generator: str = "random"
    """ generator: type of the program generator """
    instruction_set: str = "x86-64"
    """ instruction_set: ISA under test """
    instruction_categories: List[str] = []
    """ instruction_categories: list of instruction categories to use for generating programs """
    instruction_allowlist: List[str] = []
    """ instruction_allowlist: list of instructions to use for generating programs;
    combined with instruction_categories; has priority over instruction_blocklist.
    The resulting list is:
     (instructions from instruction_categories - instruction_blocklist) + instruction_allowlist """
    instruction_blocklist: List[str] = []
    """ instruction_blocklist: list of instruction that will NOT be used for generating programs;
    filters out instructions from instruction_categories, but not from instruction_allowlist.
    The resulting list is:
     (instructions from instruction_categories - instruction_blocklist) + instruction_allowlist """
    instruction_blocklist_append: List[str] = []
    """ instruction_blocklist_append: same as instruction_blocklist, but the list is added
    to the existing blocklist instead of replacing it """
    program_generator_seed: int = 0
    """ program_generator_seed: seed of the program generator """
    program_size: int = 24
    """ program_size: size of generated programs """
    avg_mem_accesses: int = 12
    """ avg_mem_accesses: average number of memory accesses in generated programs """
    min_bb_per_function: int = 1
    """ min_bb_per_function: minimal number of basic blocks per function in generated programs """
    max_bb_per_function: int = 2
    """ max_bb_per_function: maximum number of basic blocks per function in generated programs """
    min_successors_per_bb: int = 1
    """ min_bb_per_function: min. number of successors for each basic block in generated programs
    Note 1: this config option is a *hint*; it could be ignored if the instruction set does not
    have the necessary instructions to satisfy it, or if a certain number of successor is required
    for correctness"""
    max_successors_per_bb: int = 1
    """ min_bb_per_function: min. number of successors for each basic block in generated programs
    Note: this config option is a *hint*; it could be ignored if the instruction set does not
    have the necessary instructions to satisfy it, or if a certain number of successor is required
    for correctness """
    register_allowlist: List[str] = []
    """ register_allowlist: list of registers that CAN be used for generating programs;
     has higher priority than register_blocklist.
     The resulting list is: (all registers - register_blocklist) + register_allowlist """
    register_blocklist: List[str] = []
    """ register_blocklist: list of registers that will NOT be used for generating programs;
     has lower priority than register_allowlist.
     The resulting list is: (all registers - register_blocklist) + register_allowlist """
    generator_faults_allowlist: List[str] = []
    """ generator_faults_allowlist: by default, generator will produce programs that never
    trigger exceptions. This option modifies this behavior by permitting the generator to produce
    'unsafe' instruction sequences that could potentially trigger an exception. Model and executor
     will also be configured to handle these exceptions gracefully """

    # ==============================================================================================
    # Input Generator
    input_generator: str = 'random'
    """ input_generator: type of the input generator """
    input_gen_seed: int = 10
    """ input_gen_seed: input generation seed; will use a random seed if set to zero """
    input_gen_entropy_bits: int = 16
    """ input_gen_entropy_bits: entropy of the random values created by the input generator """
    inputs_per_class: int = 2
    """ inputs_per_class: number of inputs per input class """

    # ==============================================================================================
    # Contract Model
    model: str = 'x86-unicorn'
    """ model: """
    contract_execution_clause: List[str] = ["seq"]
    """ contract_execution_clause: """
    contract_observation_clause: str = 'ct'
    """ contract_observation_clause: """
    model_min_nesting: int = 1
    """ model_max_nesting: """
    model_max_nesting: int = 30
    """ model_max_nesting: """
    model_max_spec_window: int = 250
    """ model_max_spec_window: """

    # ==============================================================================================
    # Executor
    executor: str = x86_config.try_get_cpu_vendor()
    """ executor: executor type """
    executor_mode: str = 'P+P'
    """ executor_mode: hardware trace collection mode """
    executor_warmups: int = 5
    """ executor_warmups: number of warmup rounds executed before starting to collect
    hardware traces """
    executor_sample_sizes: List[int] = [10, 50, 100, 500]
    """ executor_sample_sizes: a list of sample sizes to be used during the measurements;
    the executor will first collect the hardware traces with the first sample size in the list,
    and if a violation is detected, it will try to reproduce it with all the following
    sample sizes """
    executor_filtering_repetitions: int = 10
    """ executor_filtering_repetitions: number of repetitions while filtering test cases """
    executor_taskset: int = 0
    """ executor_taskset: id of the CPU core on which the executor is running test cases """
    enable_pre_run_flush: bool = True
    """ enable_pre_run_flush: if enabled, the executor will do its best to flush
    the microarchitectural state before running test cases """

    # ==============================================================================================
    # Analyser
    analyser: str = 'chi2'
    """ analyser: type of the analyser """
    analyser_subsets_is_violation: bool = False
    """ analyser_subsets_is_violation: [only for analyser='sets' or analyser='bitmaps']
    if False, the analyser will not label hardware traces as mismatching if they form
    a subset relation """
    analyser_outliers_threshold: float = 0.1
    """ analyser_outliers_threshold: [only for analyser='sets' or analyser='bitmaps']
    analyser will ignore the htraces that appear in less then this percentage of the repetitions.
    I.e., a htrace passes the filter if it is observed at least
        (analyser_outliers_threshold * len(htrace)) times """
    analyser_stat_threshold: float = 0.5
    """ analyser_stat_threshold: [only for analyser='chi2' and analyser='mwu']
    Threshold for the statistical tests. If a pair of hardware traces has the (normalized)
    statistics below the threshold, then the traces are considered equivalent.

    Note: The threshold default value (0.5) is conservative and avoids false positives
    at cost of false negatives. For more precise results, set the threshold to a lower value.

    For the chi2 test, the threshold is the statistics / (len(htrace1) + len(htrace2))
    For the mwu test, the threshold is the p-value """

    # ==============================================================================================
    # Coverage
    coverage_type: str = 'none'
    """ coverage_type: coverage type """

    # ==============================================================================================
    # Minimizer
    minimizer_retries: int = 1
    """ minimizer_retries: number of attempts to reproduce the violation when minimizing """

    # ==============================================================================================
    # Output
    multiline_output: bool = False
    """ multiline_output: """
    logging_modes: List[str] = ["info", "stat"]
    """ logging_modes: """
    color: bool = False

    # ==============================================================================================
    # Alternatives for config options (also extended by ISA-specific config.py)
    _option_values: Dict[str, List] = {
        "fuzzer": ["basic", "architectural", "archdiff"],
        "generator": ["random"],
        "instruction_set": ["x86-64"],
        "input_generator": ["random"],
        "model": ["x86-unicorn"],
        "contract_execution_clause": [
            "seq", "no_speculation", "seq-assist", "cond", "conditional_br_misprediction", "bpas",
            "nullinj-fault", "nullinj-assist", "delayed-exception-handling", "div-zero",
            "div-overflow", "meltdown", "fault-skip", "noncanonical", "vspec-ops-div",
            "vspec-ops-memory-faults", "vspec-ops-memory-assists", "vspec-ops-gp", "vspec-all-div",
            "vspec-all-memory-faults", "vspec-all-memory-assists", "noninterference"
        ],
        "contract_observation_clause": [
            "none", "l1d", "pc", "memory", "ct", "loads+stores+pc", "ct-nonspecstore", "ctr",
            "arch", "tct", "tcto"
        ],
        "executor": ["x86-intel", "x86-amd"],
        "analyser": ["bitmaps", "sets", "mwu", "chi2"],
        "coverage_type": ["none", "model_instructions"],
        "logging_modes": [
            "info",
            "stat",
            "dbg_generator",
            "dbg_timestamp",
            "dbg_violation",
            "dbg_dump_htraces",
            "dbg_dump_ctraces",
            "dbg_dump_traces_unlimited",
            "dbg_model",
            "dbg_coverage",
            "dbg_priming",
            "dbg_executor_raw",
        ],
    }

    # ==============================================================================================
    # Internal
    _borg_shared_state: Dict = {}
    _no_generation: bool = False
    _handled_faults: List[str]  # set by ISA-specific config.py
    _generator_fault_to_fault_name: Dict[str, str]  # set by ISA-specific config.py
    _actors: OrderedDict[str, Dict]
    _actor_default: Dict
    _config_path: str = ""

    def __init__(self) -> None:
        # implementation of Borg pattern
        setattr(self, '__dict__', self._borg_shared_state)
        if not getattr(self, '_actors', None):
            self._actors = OrderedDict()

    def load(self, config_path: str, include_dir: str = "") -> None:
        self._config_path = config_path
        config_update: Dict = {}
        with open(config_path, "r") as f:
            loader = IncludeLoader(f, include_dir)
            try:
                config_update = loader.get_single_data()
            except yaml.scanner.ScannerError as e:  # type: ignore
                raise ConfigException(
                    f"Error parsing the configuration file {config_path}:\nError: {e}")
            finally:
                loader.dispose()
        self._load_from_dict(config_update)
        self._value_sanity_check()

    def _load_from_dict(self, config_update: Dict) -> None:
        # make sure to set the architecture-dependent defaults first
        if 'instruction_set' in config_update:
            self.instruction_set = config_update['instruction_set']
            self.set_to_arch_defaults()
            config_update.pop('instruction_set')

        # recursively parse the included file
        if 'file' in config_update:
            self._load_from_dict(config_update['file'])
            config_update.pop('file')

        # set the rest of the options
        for var, value in config_update.items():
            # print(f"CONF: setting {var} to {value}")
            if var == "generator_faults_allowlist":
                self.update_handled_faults_with_generator_faults(value)
                self.safe_set(var, value)
                continue
            if var == "actors":
                self.set_actor_properties(value)
                continue

            self.safe_set(var, value)

    def safe_set(self, name: str, value) -> None:
        assert name not in ["instruction_set"]

        # sanity checks
        if name[0] == "_":
            raise ConfigException(f"Attempting to set an internal configuration variable {name}.")
        if getattr(self, name, None) is None:
            raise ConfigException(f"Unknown configuration variable {name}.\n"
                                  f"It's likely a typo in the configuration file.")
        if type(self.__getattribute__(name)) != type(value):
            raise ConfigException(f"Wrong type of the configuration variable {name}.\n"
                                  f"It's likely a typo in the configuration file.")

        self._check_options(name, value)
        setattr(self, name, value)

    def _check_options(self, name: str, value) -> None:
        if name not in self._option_values:
            return
        options = self._option_values[name]

        invalid_value = None
        if isinstance(value, str):
            invalid_value = value if value not in options else None
        elif isinstance(value, List):
            for v in value:
                if v in options:
                    continue
                if isinstance(v, Dict):
                    for k in v:
                        if k not in options:
                            break
                    else:
                        continue
                invalid_value = v
                break
        else:
            raise ConfigException(f"Unexpected type of config variable {name}")

        if invalid_value:
            raise ConfigException(f"Unknown value '{invalid_value}' of config variable '{name}'\n"
                                  f"Possible options: {options}")
        return

    def _value_sanity_check(self) -> None:
        """
        Check if the configuration values make sense
        """
        if self.input_gen_entropy_bits > 32:
            raise ConfigException("input_gen_entropy_bits must be less or equal to 32 bits")
        if self.min_successors_per_bb > self.max_successors_per_bb:
            raise ConfigException("min_successors_per_bb is larger than max_successors_per_bb")

    def set_to_arch_defaults(self):
        """ Set config options according to the architecture-specific defaults """

        if self.instruction_set == "x86-64":
            config = x86_config
        else:
            raise ConfigException(f"Unknown architecture {self.instruction_set}")

        config_defaults = {}
        for c in dir(config):
            if c.startswith("__"):
                continue
            values = getattr(config, c)
            if type(values) not in [bool, int, float, str, dict, list]:
                continue
            config_defaults[c] = values

        if "_option_values" not in config_defaults:
            raise ConfigException("ISA-specific config.py must define _option_values")

        for name, value in config_defaults.items():
            if name == "generator_faults_allowlist":
                self.update_handled_faults_with_generator_faults(value)
                continue
            if name == "_actor_default":
                self._actor_default = deepcopy(value)
                self._actors = OrderedDict()
                self._actors['main'] = deepcopy(value)
                continue
            if name == "_option_values":
                for k, v in value.items():
                    self._option_values[k] = v
                continue

            setattr(self, name, value)

    def update_handled_faults_with_generator_faults(self, new: List[str]):
        for gen_fault in new:
            if not gen_fault:
                continue
            if gen_fault not in self._generator_fault_to_fault_name:
                raise ConfigException(f"Unknown generator fault {gen_fault}")
            fault = self._generator_fault_to_fault_name[gen_fault]
            if fault not in self._handled_faults:
                self._handled_faults.append(fault)

    def set_actor_properties(self, new):
        for actor_dict in new:
            name = next(iter(actor_dict))
            self._check_options("actor", actor_dict[name])
            update = {k: v for tmp_dict in actor_dict[name] for k, v in tmp_dict.items()}

            if name == "main":
                if update.get('mode', 'host') != 'host':
                    raise ConfigException("The main actor must be in 'host' mode")
                if update.get('privilege_level', 'kernel') != 'kernel':
                    raise ConfigException("The main actor must have 'kernel' privilege_level")

            if name in self._actors:
                entry = self._actors[name]
            else:
                entry = deepcopy(self._actor_default)

            for k, v in update.items():
                if k == "mode" and v not in self._option_values["actor_mode"]:
                    raise ConfigException(f"Unsupported actor mode {v}")
                if k == "privilege_level" and v not in self._option_values["actor_privilege_level"]:
                    raise ConfigException(f"Unsupported actor privilege_level {v}")

                if k == "data_properties":
                    for property_ in v:
                        for p_key, p_value in property_.items():
                            if p_key not in self._option_values["actor_data_properties"]:
                                raise ConfigException(
                                    f"Unsupported actor data_properties value {p_key}")
                            entry[k][p_key] = p_value
                    continue
                if k == "data_ept_properties":
                    if update.get('mode', 'host') != 'guest':
                        raise ConfigException("data_ept_properties can only be used in guest mode")
                    for property_ in v:
                        for p_key, p_value in property_.items():
                            if p_key not in self._option_values["actor_data_ept_properties"]:
                                raise ConfigException(
                                    f"Unsupported actor data_ept_properties value {p_key}")
                            entry[k][p_key] = p_value
                    continue
                if k == "instruction_blocklist" or k == "fault_blocklist":
                    if v:
                        entry[k].update(v)
                    continue

                entry[k] = v
            self._actors[name] = entry


CONF = Conf()
CONF.set_to_arch_defaults()
